ipam/actor.go

```diff
@@ -0,0 +1,165 @@
+package ipam
+
+import (
+	"fmt"
+	"github.com/zettio/weave/router"
+	"net"
+	"time"
+)
+
+// Start runs the allocator goroutine
+func (alloc *Allocator) Start() {
+	actionChan := make(chan func(), router.ChannelSize)
+	alloc.actionChan = actionChan
+	go alloc.actorLoop(actionChan)
+}
+
+// Actor client API
+
+// GetFor (Sync) - get IP address for container with given name
+func (alloc *Allocator) GetFor(ident string, cancelChan <-chan bool) net.IP {
+	resultChan := make(chan net.IP, 1) // len 1 so actor can send while cancel is in progress
+	alloc.actionChan <- func() {
+		if alloc.shuttingDown {
+			resultChan <- nil
+			return
+		}
+		alloc.electLeaderIfNecessary()
+		if !alloc.tryAllocateFor(ident, resultChan) {
+			alloc.pending = append(alloc.pending, pendingAllocation{resultChan, ident})
+		}
+	}
+	select {
+	case result := <-resultChan:
+		return result
+	case <-cancelChan:
+		alloc.actionChan <- func() {
+			for i, pending := range alloc.pending {
+				if pending.Ident == ident {
+					alloc.pending = append(alloc.pending[:i], alloc.pending[i+1:]...)
+					break
+				}
+			}
+		}
+		return nil
+	}
+}
+
+// Free (Sync) - release IP address for container with given name
+func (alloc *Allocator) Free(ident string, addr net.IP) error {
+	resultChan := make(chan error)
+	alloc.actionChan <- func() {
+		if alloc.removeOwned(ident, addr) {
+			resultChan <- alloc.spaceSet.Free(addr)
+		} else {
+			resultChan <- fmt.Errorf("free: %s not owned by %s", addr, ident)
+		}
+	}
+	return <-resultChan
+}
+
+// Sync.
+func (alloc *Allocator) String() string {
+	resultChan := make(chan string)
+	alloc.actionChan <- func() {
+		resultChan <- alloc.string()
+	}
+	return <-resultChan
+}
+
+// ContainerDied is provided to satisfy the updater interface; does a free underneath.  Async.
+func (alloc *Allocator) ContainerDied(ident string) error {
+	alloc.debugln("Container", ident, "died; releasing addresses")
+	alloc.actionChan <- func() {
+		for _, ip := range alloc.owned[ident] {
+			alloc.spaceSet.Free(ip)
+		}
+		delete(alloc.owned, ident)
+	}
+	return nil // this is to satisfy the ContainerObserver interface
+}
+
+// OnShutdown (Sync)
+func (alloc *Allocator) Shutdown() {
+	alloc.infof("Shutdown")
+	doneChan := make(chan bool)
```

Comment:
this looks like a one-shot signal channel, in which case make it of type `struct{}` and close it instead of sending `true`.